src/System.Private.CoreLib/shared/System/Marvin.cs

```diff
@@ -28,73 +28,194 @@ internal static partial class Marvin
         /// </summary>
         public static int ComputeHash32(ref byte data, int count, uint p0, uint p1)
         {
-            nuint ucount = (nuint)count;
-            nuint byteOffset = 0;
+            // Control flow of this method generally flows top-to-bottom, trying to
+            // minimize the number of branches taken for large (>= 8 bytes, 4 chars) inputs.
+            // If small inputs (< 8 bytes, 4 chars) are given, this jumps to a "small inputs"
+            // handler at the end of the method.
 
-            while (ucount >= 8)
+            if ((uint)count < 8)
             {
-                p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset));
-                Block(ref p0, ref p1);
+                goto InputTooSmallToEnterMainLoop;
+            }
+
+            // Main loop - read 8 bytes at a time.
+            // The block function is unrolled 2x in this loop.
+
+            uint loopCount = (uint)count / 8;
+            Debug.Assert(loopCount > 0, "Shouldn't reach this code path for small inputs.");
+
+            do
+            {
+                // Most x86 processors have two dispatch ports for reads, so we can read 2x 32-bit
+                // values in parallel. We opt for this instead of a single 64-bit read since the
+                // typical use case for Marvin32 is computing String hash codes, and the particular
+                // layout of String instances means the starting data is never 8-byte aligned when
+                // running in a 64-bit process.
+
+                p0 += Unsafe.ReadUnaligned<uint>(ref data);
+                uint nextUInt32 = Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, 4));
 
-                p0 += Unsafe.ReadUnaligned<uint>(ref Unsafe.AddByteOffset(ref data, byteOffset + 4));
+                // One block round for each of the 32-bit integers we just read, 2x rounds total.
+
+                Block(ref p0, ref p1);
+                p0 += nextUInt32;
                 Block(ref p0, ref p1);
 
-                byteOffset += 8;
-                ucount -= 8;
+                // Bump the data reference pointer and decrement the loop count.
+
+                // TODO: By decrementing the loop count by 1 every time and comparing against zero,
```

Comment:
Is there an issue tracking this?

Reply:
https://github.com/dotnet/coreclr/issues/19074 is fairly close. @tannergooding, do you know if there's a master tracking issue for making the JIT understand that some instructions set register flags and thus intermediate `test` / `cmp` instructions can be elided?

Reply:
Found a better tracking issue. See https://github.com/dotnet/coreclr/issues/7566#issuecomment-334262219.